### ISSUE TITLE
Research and analysis of current state of wazuh content

### ISSUE THREAD

**Issue by MiguelazoDS** [wazuh/wazuh#13826]
# Description
As part of the epic https://github.com/wazuh/wazuh/issues/13760 that aims to increase the effectiveness of Wazuh content, we need to understand its current state and what we need to make it a maintainable general-purpose tool capable of providing information to multiple Wazuh components that may need updated and reliable information. 

Refer to the previous analysis https://github.com/wazuh/wazuh/issues/5206 for an example for vulnerability detector, and discuss the best approach that must be followed to achieve the goal.

**Comment by pereyra-m** [comment]
# Software security feeds

## Abstract

Before the team can decide the best format for the Wazuh Unified Feed, it's important to briefly describe the way the industry manages this complex type of data.

Here is a brief and simple description of some available options that other vendors are using: OVAL, SCAP and CVRF/CSAF.
Some tools to parse and manage this type of content are provided as example.


## Introduction

Some important definitions are required before we start.

We can say that a vulnerability database is [[1]](https://en.wikipedia.org/wiki/Vulnerability_database):

`
... a platform aimed at collecting, maintaining, and disseminating information about discovered computer security vulnerabilities. The database will customarily describe the identified vulnerability, assess the potential impact on affected systems, and any workarounds or updates to mitigate the issue. A VDB will assign a unique identifier to each vulnerability cataloged such as a number (e.g. 123456) or alphanumeric designation (e.g. VDB-2020-12345).
`

For example, the National Vulnerability Database (NVD), run by the National Institute of Standards and Technology (NIST) (https://nvd.nist.gov/); and Common Vulnerabilities and Exposures (CVE) run by Mitre (https://cve.mitre.org/).

We also need to define Threat Intelligence [[2]](https://www.oreilly.com/library/view/threat-intelligence-in/9781492049302/ch01.html)

`
Threat intelligence is evidence-based knowledge, including context, mechanisms, indicators, implications and actionable advice, about an existing or emerging menace or hazard to assets that can be used to inform decisions regarding the subject’s response to that menace or hazard.
`

So we need that every single element in the Wazuh feed has its own:
- Unique identifier
- A detailed description of the elements and/or configurations (conditions) that must be present to consider certain system vulnerable to certain attack and/or undesired situation/behavior
- Extra information, external links, severity, impact, etc. to contextualize the situation. This helps to take decisions on particular cases
- The required configuration, patches, fixes and/or workaround to mitigate/solve the situation (if any)

The information of the feed comes mainly for external sources of the organization to protect (but some custom content should be allowed to be created). The verification and sanitization of this data is crucial for the success of the feed.

But we also need internal information. This should be accurate and reliable information, gathered to properly correlate it against the database for obtaining valuable and actionable data.

We don't need a simple data feed [[2]](https://www.oreilly.com/library/view/threat-intelligence-in/9781492049302/ch01.html):
 `
  A data feed is a list of indicators provided by a third party that can be correlated against internal security systems to find matches that can be acted upon.
 `

 Without the proper context, this type of feed would generate almost no value, or even create more work for the security teams.

## Different formats for software security feeds and related tools

###  Common Vulnerability Reporting Framework (CVRF) / Common Security Advisory Framework (CSAF)

- Used by CVE Mitre. Manteined by Industry Consortium for Advancement of Security on the Internet (ICASI) (http://www.icasi.org/, now FIRST https://www.first.org). See also https://oasis-open.github.io/csaf-documentation/ and https://www.oasis-open.org/committees/tc_home.php?wg_abbrev=csaf

We have parsers available (https://github.com/oasis-open/csaf-parser) to convert this schema to CSV for example.



<details>
<summary>
Example
</summary>

```xml
<?xml version="1.0" encoding="utf-8"?>
<cvrfdoc xmlns:cpe="http://cpe.mitre.org/language/2.0"
  xmlns:cvrf="http://docs.oasis-open.org/csaf/ns/csaf-cvrf/v1.2/cvrf"
  xmlns:cvrf-common="http://docs.oasis-open.org/csaf/ns/csaf-cvrf/v1.2/common"
  xmlns:cvssv2="http://scap.nist.gov/schema/cvss-v2/1.0"
  xmlns:cvssv3="https://www.first.org/cvss/cvss-v3.0.xsd"
  xmlns:dc="http://purl.org/dc/elements/1.1/"
  xmlns:prod="http://docs.oasis-open.org/csaf/ns/csaf-cvrf/v1.2/prod"
  xmlns:scap-core="http://scap.nist.gov/schema/scap-core/1.0"
  xmlns:sch="http://purl.oclc.org/dsdl/schematron"
  xmlns:vuln="http://docs.oasis-open.org/csaf/ns/csaf-cvrf/v1.2/vuln"
  xmlns:xsi=http://www.w3.org/2001/XMLSchema-instance
  xmlns="http://docs.oasis-open.org/csaf/ns/csaf-cvrf/v1.2/cvrf"
  >
  <!-- Document wide context information -->
  <DocumentTitle xml:lang="en">Red Hat Security Advisory: python-oslo-middleware security update</DocumentTitle>
  <DocumentType>Security Advisory</DocumentType>
  <DocumentPublisher Type="Vendor">
    <ContactDetails>secalert@redhat.com</ContactDetails>
    <IssuingAuthority>Red Hat Product Security</IssuingAuthority>
  </DocumentPublisher>
  <DocumentTracking>
    <Identification>
      <ID>RHSA-2017:0435</ID>
    </Identification>
    <Status>Final</Status>
    <Version>1</Version>
    <RevisionHistory>
      <Revision>
        <Number>1</Number>
        <Date>2017-03-02T21:13:00Z</Date>
        <Description>Current version</Description>
      </Revision>
    </RevisionHistory>
    <InitialReleaseDate>2017-03-02T21:13:00Z</InitialReleaseDate>
    <CurrentReleaseDate>2017-03-02T21:13:00Z</CurrentReleaseDate>
    <Generator>
      <Engine>Red Hat rhsa-to-cvrf 2.0</Engine>
      <Date>2017-03-04T05:06:05Z</Date>
    </Generator>
  </DocumentTracking>
  <DocumentNotes>
    <Note Title="Topic" Type="Summary" Ordinal="1" xml:lang="en">
An update for python-oslo-middleware is now available for Red Hat OpenStack Platform 9.0 (Mitaka).

Red Hat Product Security has rated this update as having a security impact of Moderate. A Common Vulnerability Scoring System (CVSS) base score, which gives a detailed severity rating, is available for each vulnerability from the CVE link(s) in the References section.    </Note>
    <Note Title="Details" Type="General" Ordinal="2" xml:lang="en">
The OpenStack Oslo Middleware library provides components that can be injected into WSGI pipelines to intercept request and response flows. The base class can be enhanced with functionality like adding or updating HTTP headers, or to offer support for limiting size or connections.

Security Fix(es):

* An information-disclosure flaw was found in oslo.middleware. Software using the CatchError class could include sensitive values in a traceback's error message. System users could exploit this flaw to obtain sensitive information from OpenStack component error logs (for example, keystone tokens). (CVE-2017-2592)

Red Hat would like to thank the OpenStack project for reporting this issue. Upstream acknowledges Divya K Konoor (IBM) as the original reporter.    </Note>
    <Note Title="Terms of Use" Ordinal="3" Type="Legal Disclaimer" xml:lang="en">Please see https://www.redhat.com/footer/terms-of-use.html</Note>
  </DocumentNotes>
  <DocumentDistribution xml:lang="en">Copyright © 2017 Red Hat, Inc. All rights reserved.</DocumentDistribution>
  <AggregateSeverity Namespace="https://access.redhat.com/security/updates/classification/">Moderate</AggregateSeverity>
  <DocumentReferences>
    <Reference Type="Self">
      <URL>https://rhn.redhat.com/errata/RHSA-2017-0435.html</URL>
      <Description>https://rhn.redhat.com/errata/RHSA-2017-0435.html</Description>
    </Reference>
    <Reference>
      <URL>https://access.redhat.com/security/updates/classification/#moderate</URL>
      <Description>https://access.redhat.com/security/updates/classification/#moderate</Description>
    </Reference>
  </DocumentReferences>
  <!-- Product tree section -->
  <prod:ProductTree xmlns="http://docs.oasis-open.org/csaf/ns/csaf-cvrf/v1.2/prod">
    <Branch Type="Product Family" Name="Red Hat Enterprise Linux OpenStack Platform">
      <Branch Type="Product Name" Name="Red Hat OpenStack Platform 9.0">
        <FullProductName ProductID="7Server-RH7-RHOS-9.0">Red Hat OpenStack Platform 9.0</FullProductName>
      </Branch>
    </Branch>
    <Branch Type="Product Version" Name="python-oslo-middleware-3.7.0-2.el7ost">
      <FullProductName ProductID="python-oslo-middleware-3.7.0-2.el7ost">python-oslo-middleware-3.7.0-2.el7ost.src.rpm</prod:FullProductName>
    </Branch>
    <Relationship ProductReference="python-oslo-middleware-3.7.0-2.el7ost"
      RelationType="Default Component Of" RelatesToProductReference="7Server-RH7-RHOS-9.0">
      <FullProductName ProductID="7Server-RH7-RHOS-9.0:python-oslo-middleware-3.7.0-2.el7ost">python-oslo-middleware-3.7.0-2.el7ost as a component of Red Hat OpenStack Platform 9.0</FullProductName>
    </Relationship>
  </ProductTree>
  <!-- Vulnerability section -->
  <vuln:Vulnerability Ordinal="1"
   xmlns="http://docs.oasis-open.org/csaf/ns/csaf-cvrf/v1.2/vuln">
    <Notes>
      <Note Title="Vulnerability Description" Type="General" Ordinal="1" xml:lang="en">An information-disclosure flaw was found in oslo.middleware. Software using the CatchError class could include sensitive values in a traceback's error message. System users could exploit this flaw to obtain sensitive information from OpenStack component error logs (for example, keystone tokens). </Note>
    </Notes>
    <DiscoveryDate>2017-01-18T00:00:00Z</DiscoveryDate>
    <ReleaseDate>2017-01-26T00:00:00Z</ReleaseDate>
    <Involvements>
      <Involvement Party="Vendor" Status="Completed"/>
    </Involvements>
    <CVE>CVE-2017-2592</CVE>
    <ProductStatuses>
      <Status Type="Fixed">
        <ProductID>7Server-RH7-RHOS-9.0:python-oslo-middleware-3.7.0-2.el7ost</ProductID>
      </Status>
    </ProductStatuses>
    <Threats>
      <Threat Type="Impact">
        <Description>Moderate</Description>
      </Threat>
    </Threats>

    <Remediations>
      <Remediation Type="Vendor Fix">
        <Description xml:lang="en">
For details on how to apply this update, which includes the changes described in this advisory, refer to:

https://access.redhat.com/articles/11258    </Description>
        <URL>https://rhn.redhat.com/errata/RHSA-2017-0435.html</URL>
      </Remediation>
    </Remediations>
    <References>
      <Reference>
        <URL>https://access.redhat.com/security/cve/CVE-2017-2592</URL>
        <Description>CVE-2017-2592</vuln:Description>
      </Reference>
      <Reference>
        <URL>https://bugzilla.redhat.com/show_bug.cgi?id=1414698</URL>
        <Description>bz#1414698: CVE-2017-2592 python-oslo-middleware: CatchErrors leaks sensitive values into error logs</Description>
      </Reference>
    </References>
    <Acknowledgments>
      <Acknowledgment>
        <Description>Red Hat would like to thank the OpenStack project for reporting this issue. Upstream acknowledges Divya K Konoor (IBM) as the original reporter.</Description>
      </Acknowledgment>
    </Acknowledgments>
  </Vulnerability>
  <!-- No more elements to follow -->
</cvrfdoc>
```

</details>

### Github: OASIS
- Source: https://github.com/oasis-open/csaf-parser. The examples fail (https://github.com/oasis-open/csaf-parser/issues/23).
- Source: https://github.com/oasis-tcs/csaf. The are scripts without examples to run.
---

###  Security Content Automation Protocol (SCAP)

- Used by NVD NIST

<details>
<summary>
XML structure
</summary>

```xml
This XML file does not appear to have any style information associated with it. The document tree is shown below.
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:ds="http://scap.nist.gov/schema/scap/source/1.2" xmlns:xccdf="http://checklists.nist.gov/xccdf/1.2" xmlns:oval="http://oval.mitre.org/XMLSchema/oval-definitions-5" xmlns:cpe-dict="http://cpe.mitre.org/dictionary/2.0" xmlns:ocil="http://scap.nist.gov/schema/ocil/2.0" xmlns:dsig="http://www.w3.org/2000/09/xmldsig#" xmlns:cat="urn:oasis:names:tc:entity:xmlns:xml:catalog" xmlns:xlink="http://www.w3.org/1999/xlink" targetNamespace="http://scap.nist.gov/schema/scap/source/1.2" elementFormDefault="qualified" attributeFormDefault="unqualified" version="1.2">
<script/>
<script/>
<xs:annotation>
<xs:appinfo>
<schema>SCAP 1.2 Source Data Stream Collection</schema>
<author>Adam Halbardier, David Waltermire</author>
<version>1.2</version>
<date>2011-10-10</date>
</xs:appinfo>
</xs:annotation>
<xs:import namespace="http://checklists.nist.gov/xccdf/1.2" schemaLocation="https://scap.nist.gov/schema/xccdf/1.2/xccdf_1.2.xsd"/>
<xs:import namespace="http://oval.mitre.org/XMLSchema/oval-definitions-5" schemaLocation="http://oval.mitre.org/language/version5.10.1/ovaldefinition/complete/oval-definitions-schema.xsd"/>
<xs:import namespace="http://cpe.mitre.org/dictionary/2.0" schemaLocation="https://scap.nist.gov/schema/cpe/2.3/cpe-dictionary_2.3.xsd"/>
<xs:import namespace="http://scap.nist.gov/schema/ocil/2.0" schemaLocation="https://scap.nist.gov/schema/ocil/2.0/ocil-2.0.xsd"/>
<xs:import namespace="http://www.w3.org/2000/09/xmldsig#" schemaLocation="https://www.w3.org/TR/xmldsig-core/xmldsig-core-schema.xsd"/>
<xs:import namespace="urn:oasis:names:tc:entity:xmlns:xml:catalog" schemaLocation="http://www.oasis-open.org/committees/entity/release/1.1/catalog.xsd"/>
<xs:import namespace="http://www.w3.org/1999/xlink" schemaLocation="https://www.w3.org/1999/xlink.xsd"/>
<xs:element name="data-stream-collection">
<xs:annotation>
<xs:documentation>Holds a collection of data streams and components.</xs:documentation>
</xs:annotation>
<xs:complexType>
<xs:sequence>
<xs:element maxOccurs="unbounded" ref="ds:data-stream"/>
<xs:element maxOccurs="unbounded" ref="ds:component"/>
<xs:element maxOccurs="unbounded" minOccurs="0" ref="ds:extended-component"/>
<xs:element maxOccurs="unbounded" minOccurs="0" ref="dsig:Signature">
<xs:annotation>
<xs:documentation>A digital signature of a data stream.</xs:documentation>
</xs:annotation>
</xs:element>
</xs:sequence>
<xs:attribute name="id" use="required">
<xs:annotation>
<xs:documentation>This MUST be a globally unique ID.</xs:documentation>
</xs:annotation>
<xs:simpleType>
<xs:restriction base="xs:ID">
<xs:pattern value="scap_[^_]+_collection_.+"/>
</xs:restriction>
</xs:simpleType>
</xs:attribute>
<xs:attribute name="schematron-version" type="xs:token" use="required">
<xs:annotation>
<xs:documentation>The version of the requirements Schematron ruleset to which the instance conforms.</xs:documentation>
</xs:annotation>
</xs:attribute>
</xs:complexType>
</xs:element>
<xs:element name="data-stream">
<xs:annotation>
<xs:documentation>An SCAP data stream containing pointers to all of the components composing the data stream.</xs:documentation>
</xs:annotation>
<xs:complexType>
<xs:sequence>
<xs:element name="dictionaries" type="ds:refListType" minOccurs="0">
<xs:annotation>
<xs:documentation>Holds pointers to dictionary components.</xs:documentation>
</xs:annotation>
</xs:element>
<xs:element name="checklists" type="ds:refListType" minOccurs="0">
<xs:annotation>
<xs:documentation>Holds pointers to checklist components.</xs:documentation>
</xs:annotation>
</xs:element>
<xs:element name="checks" type="ds:refListType">
<xs:annotation>
<xs:documentation>Holds pointers to check components.</xs:documentation>
</xs:annotation>
</xs:element>
<xs:element minOccurs="0" name="extended-components" type="ds:refListType">
<xs:annotation>
<xs:documentation>Holds pointers to non-standard SCAP components captured as extended-component elements.</xs:documentation>
</xs:annotation>
</xs:element>
</xs:sequence>
<xs:attribute name="id" use="required">
<xs:annotation>
<xs:documentation>This MUST be a globally unique ID.</xs:documentation>
</xs:annotation>
<xs:simpleType>
<xs:restriction base="xs:ID">
<xs:pattern value="scap_[^_]+_datastream_.+"/>
</xs:restriction>
</xs:simpleType>
</xs:attribute>
<xs:attribute name="use-case" use="required">
<xs:annotation>
<xs:documentation>The SCAP capability being expressed by this data stream. The type is expressed to allow for future use of this schema while indicating the currently acceptable values.</xs:documentation>
</xs:annotation>
<xs:simpleType>
<xs:union memberTypes="ds:useCaseType xs:token"/>
</xs:simpleType>
</xs:attribute>
<xs:attribute name="scap-version" use="required">
<xs:annotation>
<xs:documentation>The version of SCAP expressed by this data stream. The type is expressed to allow for future use of this schema while indicating the currently acceptable values.</xs:documentation>
</xs:annotation>
<xs:simpleType>
<xs:union memberTypes="ds:scapVersionType xs:token"/>
</xs:simpleType>
</xs:attribute>
<xs:attribute name="timestamp" type="xs:dateTime">
<xs:annotation>
<xs:documentation>The time when the data stream was created or last modified.</xs:documentation>
</xs:annotation>
</xs:attribute>
</xs:complexType>
</xs:element>
<xs:element name="component">
<xs:annotation>
<xs:documentation>A component that is used by an SCAP data stream.</xs:documentation>
</xs:annotation>
<xs:complexType>
<xs:choice>
<xs:element ref="xccdf:Benchmark"/>
<xs:element ref="oval:oval_definitions"/>
<xs:element ref="ocil:ocil"/>
<xs:element ref="cpe-dict:cpe-list"/>
<xs:element ref="xccdf:Tailoring"/>
</xs:choice>
<xs:attribute name="id" use="required">
<xs:annotation>
<xs:documentation>This MUST be a globally unique ID.</xs:documentation>
</xs:annotation>
<xs:simpleType>
<xs:restriction base="xs:ID">
<xs:pattern value="scap_[^_]+_comp_.+"/>
</xs:restriction>
</xs:simpleType>
</xs:attribute>
<xs:attribute name="timestamp" type="xs:dateTime" use="required">
<xs:annotation>
<xs:documentation>The time when the component was created or last modified.</xs:documentation>
</xs:annotation>
</xs:attribute>
</xs:complexType>
<xs:unique name="cpeDictionaryUnique">
<xs:selector xpath="cpe-dict:cpe-list/cpe-dict:cpe-item"/>
<xs:field xpath="@name"/>
</xs:unique>
</xs:element>
<xs:element name="extended-component">
<xs:annotation>
<xs:documentation>A component that holds non-standard SCAP content.</xs:documentation>
</xs:annotation>
<xs:complexType>
<xs:sequence>
<xs:any namespace="##other" processContents="lax"/>
</xs:sequence>
<xs:attribute name="id" use="required">
<xs:annotation>
<xs:documentation>This MUST be a globally unique ID.</xs:documentation>
</xs:annotation>
<xs:simpleType>
<xs:restriction base="xs:ID">
<xs:pattern value="scap_[^_]+_ecomp_.+"/>
</xs:restriction>
</xs:simpleType>
</xs:attribute>
<xs:attribute name="timestamp" type="xs:dateTime" use="required">
<xs:annotation>
<xs:documentation>The time when the component was created or last modified.</xs:documentation>
</xs:annotation>
</xs:attribute>
</xs:complexType>
</xs:element>
<xs:element name="component-ref">
<xs:annotation>
<xs:documentation>An XLink element that points to a component.</xs:documentation>
</xs:annotation>
<xs:complexType>
<xs:sequence>
<xs:element minOccurs="0" ref="cat:catalog"/>
</xs:sequence>
<xs:attribute name="id" use="required">
<xs:annotation>
<xs:documentation>This MUST be a globally unique ID.</xs:documentation>
</xs:annotation>
<xs:simpleType>
<xs:restriction base="xs:ID">
<xs:pattern value="scap_[^_]+_cref_.+"/>
</xs:restriction>
</xs:simpleType>
</xs:attribute>
<xs:attribute fixed="simple" ref="xlink:type"/>
<xs:attribute use="required" ref="xlink:href"/>
</xs:complexType>
</xs:element>
<xs:simpleType name="useCaseType">
<xs:restriction base="xs:token">
<xs:enumeration value="CONFIGURATION"/>
<xs:enumeration value="VULNERABILITY"/>
<xs:enumeration value="INVENTORY"/>
<xs:enumeration value="OTHER"/>
</xs:restriction>
</xs:simpleType>
<xs:simpleType name="scapVersionType">
<xs:restriction base="xs:token">
<xs:enumeration value="1.0"/>
<xs:enumeration value="1.1"/>
<xs:enumeration value="1.2"/>
<xs:enumeration value="1.3"/>
</xs:restriction>
</xs:simpleType>
<xs:complexType name="contentSourceType">
<xs:attribute name="href" type="xs:anyURI" use="required"/>
</xs:complexType>
<xs:complexType name="refListType">
<xs:sequence>
<xs:element maxOccurs="unbounded" ref="ds:component-ref"/>
</xs:sequence>
</xs:complexType>
</xs:schema>
```

</details>


<details>
<summary>
An example JSON got from NVD
</summary>

```json
{
  "CVE_data_type" : "CVE",
  "CVE_data_format" : "MITRE",
  "CVE_data_version" : "4.0",
  "CVE_data_numberOfCVEs" : "6984",
  "CVE_data_timestamp" : "2022-06-13T07:00Z",
  "CVE_Items" : [ {
    "cve" : {
      "data_type" : "CVE",
      "data_format" : "MITRE",
      "data_version" : "4.0",
      "CVE_data_meta" : {
        "ID" : "CVE-2022-0204",
        "ASSIGNER" : "secalert@redhat.com"
      },
      "problemtype" : {
        "problemtype_data" : [ {
          "description" : [ {
            "lang" : "en",
            "value" : "CWE-787"
          } ]
        } ]
      },
      "references" : {
        "reference_data" : [ {
          "url" : "https://bugzilla.redhat.com/show_bug.cgi?id=2039807",
          "name" : "https://bugzilla.redhat.com/show_bug.cgi?id=2039807",
          "refsource" : "MISC",
          "tags" : [ "Issue Tracking", "Patch", "Third Party Advisory" ]
        }, {
          "url" : "https://github.com/bluez/bluez/security/advisories/GHSA-479m-xcq5-9g2q",
          "name" : "https://github.com/bluez/bluez/security/advisories/GHSA-479m-xcq5-9g2q",
          "refsource" : "MISC",
          "tags" : [ "Exploit", "Third Party Advisory" ]
        }, {
          "url" : "https://github.com/bluez/bluez/commit/591c546c536b42bef696d027f64aa22434f8c3f0",
          "name" : "https://github.com/bluez/bluez/commit/591c546c536b42bef696d027f64aa22434f8c3f0",
          "refsource" : "MISC",
          "tags" : [ "Patch", "Third Party Advisory" ]
        } ]
      },
      "description" : {
        "description_data" : [ {
          "lang" : "en",
          "value" : "A heap overflow vulnerability was found in bluez in versions prior to 5.63. An attacker with local network access could pass specially crafted files causing an application to halt or crash, leading to a denial of service."
        } ]
      }
    },
    "configurations" : {
      "CVE_data_version" : "4.0",
      "nodes" : [ {
        "operator" : "OR",
        "children" : [ ],
        "cpe_match" : [ {
          "vulnerable" : true,
          "cpe23Uri" : "cpe:2.3:a:bluez:bluez:*:*:*:*:*:*:*:*",
          "versionEndExcluding" : "5.63",
          "cpe_name" : [ ]
        } ]
      }, {
        "operator" : "OR",
        "children" : [ ],
        "cpe_match" : [ {
          "vulnerable" : true,
          "cpe23Uri" : "cpe:2.3:o:fedoraproject:fedora:35:*:*:*:*:*:*:*",
          "cpe_name" : [ ]
        } ]
      } ]
    },
    "impact" : {
      "baseMetricV3" : {
        "cvssV3" : {
          "version" : "3.1",
          "vectorString" : "CVSS:3.1/AV:A/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H",
          "attackVector" : "ADJACENT_NETWORK",
          "attackComplexity" : "LOW",
          "privilegesRequired" : "NONE",
          "userInteraction" : "NONE",
          "scope" : "UNCHANGED",
          "confidentialityImpact" : "HIGH",
          "integrityImpact" : "HIGH",
          "availabilityImpact" : "HIGH",
          "baseScore" : 8.8,
          "baseSeverity" : "HIGH"
        },
        "exploitabilityScore" : 2.8,
        "impactScore" : 5.9
      },
      "baseMetricV2" : {
        "cvssV2" : {
          "version" : "2.0",
          "vectorString" : "AV:A/AC:L/Au:N/C:P/I:P/A:P",
          "accessVector" : "ADJACENT_NETWORK",
          "accessComplexity" : "LOW",
          "authentication" : "NONE",
          "confidentialityImpact" : "PARTIAL",
          "integrityImpact" : "PARTIAL",
          "availabilityImpact" : "PARTIAL",
          "baseScore" : 5.8
        },
        "severity" : "MEDIUM",
        "exploitabilityScore" : 6.5,
        "impactScore" : 6.4,
        "acInsufInfo" : false,
        "obtainAllPrivilege" : false,
        "obtainUserPrivilege" : false,
        "obtainOtherPrivilege" : false,
        "userInteractionRequired" : false
      }
    },
    "publishedDate" : "2022-03-10T17:44Z",
    "lastModifiedDate" : "2022-04-26T17:01Z"
  },
  ...
```

</details>

### Github: OpenSCAP

- Source: https://github.com/OpenSCAP
- The command line can be used for content validation

---

### Open Vulnerability and Assessment Language (OVAL)

- OVAL (https://oval.mitre.org/) has transitioned to the Center for Internet Security (CIS) ( https://oval.cisecurity.org/). The MITRE OVAL website is in "Archive" status.


<details>
<summary>
An XML example
</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<oval_definitions
	xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5"
	xmlns:oval="http://oval.mitre.org/XMLSchema/oval-common-5"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://oval.mitre.org/XMLSchema/oval-common-5 oval-common-schema.xsd http://oval.mitre.org/XMLSchema/oval-definitions-5 oval-definitions-schema.xsd">

	<generator>
		<oval:product_name>CIS OVAL Repository</oval:product_name>
		<oval:product_version>0.1</oval:product_version>
		<oval:schema_version>5.11.2</oval:schema_version>
		<oval:timestamp>2022-06-13T08:25:35</oval:timestamp>
	</generator>

	<definitions>		<definition xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5" class="vulnerability" id="oval:org.cisecurity:def:2183" version="9">
		  <metadata>
		    <title>Vulnerability in Oracle MySQL 5.6.35 and earlier and 5.7.17 and earlier – CVE-2017-3452</title>
		    <affected family="windows">
		      <platform>Microsoft Windows Vista</platform>
		      <platform>Microsoft Windows 7</platform>
		      <platform>Microsoft Windows 8</platform>
		      <platform>Microsoft Windows 8.1</platform>
		      <platform>Microsoft Windows 10</platform>
		      <platform>Microsoft Windows Server 2008</platform>
		      <platform>Microsoft Windows Server 2008 R2</platform>
		      <platform>Microsoft Windows Server 2012</platform>
		      <platform>Microsoft Windows Server 2012 R2</platform>
		      <platform>Microsoft Windows Server 2016</platform>
		      <product>MySQL Server 5.6</product>
		    </affected>
		    <reference ref_id="CVE-2017-3452" ref_url="https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2017-3452" source="CVE" />
		    <description>Vulnerability in the MySQL Server component of Oracle MySQL (subcomponent: Server: Optimizer). Supported versions that are affected are 5.6.35 and earlier. Easily "exploitable" vulnerability allows low privileged attacker with network access via multiple protocols to compromise MySQL Server. Successful attacks of this vulnerability can result in unauthorized ability to cause a hang or frequently repeatable crash (complete DOS) of MySQL Server. CVSS 3.0 Base Score 6.5 (Availability impacts). CVSS Vector: (CVSS:3.0/AV:N/AC:L/PR:L/UI:N/S:U/C:N/I:N/A:H).</description>
		    <oval_repository>
		      <dates>
		        <submitted date="2017-04-26T00:00:00+08:00">
		          <contributor organization="DTCC">Jeff Albert</contributor>
		        </submitted>
		        <status_change date="2017-04-28T12:00:00.000-05:00">DRAFT</status_change>
		        <status_change date="2017-05-12T11:00:00.000-05:00">INTERIM</status_change>
		        <status_change date="2017-05-26T10:00:00.000-05:00">ACCEPTED</status_change>
		      </dates>
		      <status>ACCEPTED</status>
		      <min_schema_version>5.10</min_schema_version>
		    </oval_repository>
		  </metadata>
		  <criteria comment="Check for installation of MySQL Server 5.6 + vulnerable version" operator="AND">
		    <extend_definition comment="MySQL 5.6 is installed" definition_ref="oval:org.mitre.oval:def:26414" />
		    <criterion comment="Check if MySQL Server 5.6 version is less than or equal 5.6.35" test_ref="oval:org.cisecurity:tst:3002" />
		  </criteria>
		</definition>

        ...
```

</details>

### Github: CISecurity/OVALRepo

- Source: https://github.com/CISecurity/OVALRepo/tree/master/scripts

- To test the scripts, first I installed the dependencies in https://github.com/CISecurity/OVALRepo/blob/master/scripts/README.scripting.setup.md
- There are many examples related to content validation according to schemas that weren't tested

<details>
<Summary>
Windows definitions generated
</Summary>

```python
# Compile all Windows vulnerability definitions into all.windows.vulnerability.xml
$ python3 build_oval_definitions_file.py -o all.windows.vulnerability.xml --family windows --class vulnerability

INFO: Rebuilding oval_definitions index completed (36735 definitions)
INFO: Found 13614 matching OVAL definitions
INFO: Finding downstream OVAL ids for all definitions
INFO: Rebuilding oval_elements index completed (269353 elements)
INFO: Found 65842 downstream OVAL ids
INFO: Finding paths for 79456 OVAL elements
INFO: Generating OVAL definition file with 79456 elements
INFO: Writing OVAL definitions to all.windows.vulnerability.xml
INFO: Completed in 00:01:57!

```
</details>

<details>
<Summary>
UNIX definitions for a specific CVE generated
</Summary>

```python
# Compile all UNIX definitions related to CVE-2014-6509 into CVE.2014.6509.xml
$ python3 build_oval_definitions_file.py -o CVE.2014.6509.xml --family unix --reference_id "CVE-2014-6509"

INFO: Found 1 matching OVAL definitions
INFO: Finding downstream OVAL ids for all definitions
INFO: Found 13 downstream OVAL ids
INFO: Finding paths for 14 OVAL elements
INFO: Generating OVAL definition file with 14 elements
INFO: Writing OVAL definitions to CVE.2014.6509.xml
INFO: Completed in 00:00:00!
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<oval_definitions
	xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5"
	xmlns:oval="http://oval.mitre.org/XMLSchema/oval-common-5"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://oval.mitre.org/XMLSchema/oval-common-5 oval-common-schema.xsd http://oval.mitre.org/XMLSchema/oval-definitions-5 oval-definitions-schema.xsd">

	<generator>
		<oval:product_name>CIS OVAL Repository</oval:product_name>
		<oval:product_version>0.1</oval:product_version>
		<oval:schema_version>5.11.2</oval:schema_version>
		<oval:timestamp>2022-06-13T19:59:30</oval:timestamp>
	</generator>

	<definitions>		<definition xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5" class="vulnerability" id="oval:org.mitre.oval:def:28502" version="37">
		  <metadata>
		    <title>Critical Patch Update January 2015</title>
		    <affected family="unix">
		      <platform>Sun Solaris 10</platform>
		    </affected>
		    <reference ref_id="CVE-2014-6509" ref_url="https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2014-6509" source="CVE" />
		    <description>Unspecified vulnerability in Oracle Solaris 10 allows local users to affect availability via unknown vectors related to Kernel.</description>
		    <oval_repository>
		      <dates>
		        <submitted date="2015-01-29T11:43:28.000-05:00">
		          <contributor organization="Hewlett-Packard">Puneeth</contributor>
		        </submitted>
		        <status_change date="2015-02-03T13:38:43.811-05:00">DRAFT</status_change>
		        <status_change date="2015-02-23T04:01:25.284-05:00">INTERIM</status_change>
		        <status_change date="2015-03-16T04:01:53.765-04:00">ACCEPTED</status_change>
		      </dates>
		      <status>ACCEPTED</status>
		      <min_schema_version>5.10</min_schema_version>
		    </oval_repository>
		  </metadata>
		  <criteria operator="OR">
		    <criteria comment="Patch 148403-01 or later installed" operator="AND">
		      <extend_definition comment="Solaris 10 (SPARC) is installed" definition_ref="oval:org.mitre.oval:def:1440" />
		      <criterion comment="Patch 148403-01 or later installed" negate="true" test_ref="oval:org.mitre.oval:tst:137119" />
		    </criteria>
		    <criteria comment="Patch 148404-01 or later installed" operator="AND">
		      <extend_definition comment="Solaris 10 (x86) is installed" definition_ref="oval:org.mitre.oval:def:1926" />
		      <criterion comment="Patch 148404-01 or later installed" negate="true" test_ref="oval:org.mitre.oval:tst:137600" />
		    </criteria>
		  </criteria>
		</definition>

		<definition xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5" class="inventory" id="oval:org.mitre.oval:def:1926" version="24">
		  <metadata>
		    <title>Solaris 10 (x86) is installed</title>
		    <affected family="unix">
		      <platform>Sun Solaris 10</platform>
		    </affected>
		    <reference ref_id="cpe:/o:sun:sunos:5.10::ix86" source="CPE" />
		    <description>The operating system installed on the system is Sun Solaris 10 for x86.</description>
		    <oval_repository>
		      <dates>
		        <submitted date="2007-06-15T12:00:00.000-04:00">
		          <contributor organization="The MITRE Corporation">Jonathan Baker</contributor>
		        </submitted>
		        <status_change date="2007-06-15T12:20:00.000-04:00">DRAFT</status_change>
		        <status_change date="2007-07-03T21:53:53.007-04:00">INTERIM</status_change>
		        <status_change date="2007-07-18T15:57:51.357-04:00">ACCEPTED</status_change>
		      </dates>
		      <status>ACCEPTED</status>
		      <min_schema_version>5.10</min_schema_version>
		    </oval_repository>
		  </metadata>
		  <criteria>
		    <criterion comment="Solaris 10 Installed" test_ref="oval:org.mitre.oval:tst:3680" />
		    <criterion comment="ix86 architecture" test_ref="oval:org.mitre.oval:tst:3912" />
		  </criteria>
		</definition>

		<definition xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5" class="inventory" id="oval:org.mitre.oval:def:1440" version="24">
		  <metadata>
		    <title>Solaris 10 (SPARC) is installed</title>
		    <affected family="unix">
		      <platform>Sun Solaris 10</platform>
		    </affected>
		    <reference ref_id="cpe:/o:sun:sunos:5.10::sparc" source="CPE" />
		    <description>The operating system installed on the system is Sun Solaris 10 for SPARC.</description>
		    <oval_repository>
		      <dates>
		        <submitted date="2007-06-15T12:00:00.000-04:00">
		          <contributor organization="The MITRE Corporation">Jonathan Baker</contributor>
		        </submitted>
		        <status_change date="2007-06-15T12:20:00.000-04:00">DRAFT</status_change>
		        <status_change date="2007-07-03T21:53:52.343-04:00">INTERIM</status_change>
		        <status_change date="2007-07-18T15:57:49.640-04:00">ACCEPTED</status_change>
		      </dates>
		      <status>ACCEPTED</status>
		      <min_schema_version>5.10</min_schema_version>
		    </oval_repository>
		  </metadata>
		  <criteria>
		    <criterion comment="Solaris 10 Installed" test_ref="oval:org.mitre.oval:tst:3680" />
		    <criterion comment="sparc architecture" test_ref="oval:org.mitre.oval:tst:3237" />
		  </criteria>
		</definition>

	</definitions>

	<tests>		<patch54_test xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#solaris" check="at least one" check_existence="at_least_one_exists" comment="Patch 148404-01 or later installed" id="oval:org.mitre.oval:tst:137600" version="1">
		  <object object_ref="oval:org.mitre.oval:obj:43021" />
		</patch54_test>

		<patch54_test xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#solaris" check="at least one" check_existence="at_least_one_exists" comment="Patch 148403-01 or later installed" id="oval:org.mitre.oval:tst:137119" version="1">
		  <object object_ref="oval:org.mitre.oval:obj:43124" />
		</patch54_test>

		<uname_test xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#unix" check="at least one" check_existence="at_least_one_exists" comment="Solaris 10 Installed" id="oval:org.mitre.oval:tst:3680" version="11">
		  <object object_ref="oval:org.mitre.oval:obj:2759" />
		  <state state_ref="oval:org.mitre.oval:ste:3597" />
		</uname_test>

		<uname_test xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#unix" check="at least one" check_existence="at_least_one_exists" comment="sparc architecture" id="oval:org.mitre.oval:tst:3237" version="11">
		  <object object_ref="oval:org.mitre.oval:obj:2759" />
		  <state state_ref="oval:org.mitre.oval:ste:3478" />
		</uname_test>

		<uname_test xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#unix" check="at least one" check_existence="at_least_one_exists" comment="ix86 architecture" id="oval:org.mitre.oval:tst:3912" version="11">
		  <object object_ref="oval:org.mitre.oval:obj:2759" />
		  <state state_ref="oval:org.mitre.oval:ste:3443" />
		</uname_test>

	</tests>

	<objects>		<ns0:uname_object xmlns:ns0="http://oval.mitre.org/XMLSchema/oval-definitions-5#unix" comment="The single uname object." id="oval:org.mitre.oval:obj:2759" version="1" />




		<patch54_object xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#solaris" id="oval:org.mitre.oval:obj:43021" version="1">
		  <behaviors supersedence="true" />
		  <base datatype="int">148404</base>
		  <version datatype="int">01</version>
		</patch54_object>

		<patch54_object xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#solaris" id="oval:org.mitre.oval:obj:43124" version="1">
		  <behaviors supersedence="true" />
		  <base datatype="int">148403</base>
		  <version datatype="int">01</version>
		</patch54_object>

	</objects>

	<states>		<uname_state xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#unix" comment="os release is 5.10" id="oval:org.mitre.oval:ste:3597" version="2">
		  <os_release>5.10</os_release>
		</uname_state>

		<uname_state xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#unix" comment="processor type is SPARC" id="oval:org.mitre.oval:ste:3478" version="2">
		  <processor_type operation="pattern match">[Ss][Pp][Aa][Rr][Cc]</processor_type>
		</uname_state>

		<uname_state xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#unix" comment="processor type is ix86" id="oval:org.mitre.oval:ste:3443" version="2">
		  <processor_type operation="pattern match">^i.*86</processor_type>
		</uname_state>

	</states>

</oval_definitions>
```
</details>


---

## Proof of concept for Wazuh

OVAL was selected for this purpose, because there is plenty information available. Also it seems to be really generic [[3]](https://ovalproject.github.io/getting-started/tutorial/):

```
 In OVAL, there are five kinds of definition classes:

1- Compliance: Checks whether an endpoint is compliant with a specific policy.
2- Inventory: Checks whether specific software is installed on the endpoint.
3- Miscellaneous: OVAL Definitions that do not fall into one of the other defined classes.
4- Patch: Checks whether a patch needs to be installed on an endpoint.
5- Vulnerability: Checks whether an endpoint is vulnerable.
```

Also, when content is created some basic tests must be done [[3]](https://ovalproject.github.io/getting-started/tutorial/):
```
Testing is an important part of the OVAL content development process. The aim of content testing is to evaluate the following aspects of an OVAL Definition:

Validity. The definition must be conformant to the OVAL Language specification. In other words, it must be valid OVAL content.

Interoperability. The definition must be interoperable across different implementations of interpreters and other OVAL tools.

Accuracy. The definition must make accurate evaluations of the endpoints it was written to address.

Efficiency. It must be computationally practical to evaluate an endpoint using the definition.

```

Here we have a guide to validate the created content https://ovalproject.github.io/documentation/validation/.

There are many tools for XML content creation and editing. Here, we have tested QXMLEdit https://github.com/lbellonda/qxmledit.

It has XSD Schema validation, visual editing, etc.

![xml2](https://user-images.githubusercontent.com/65046601/173454784-e750b395-9595-48a9-b2ce-87f55a821461.png)
![xml1](https://user-images.githubusercontent.com/65046601/173454812-c3f7e333-8bb5-48cf-9783-864a71665191.png)
![xml3](https://user-images.githubusercontent.com/65046601/173455133-6a58b661-2281-42df-94fe-4fbc9d1ddff3.png)


## Conclusion

It is possible to generate content for Wazuh in a common language like OVAL:
- It covers many different applications
- Many security vendors are using it
- There are many tools and documentation to create, validate and modify these XML like formats.

## References

- [1] https://en.wikipedia.org/wiki/Vulnerability_database
- [2] https://www.oreilly.com/library/view/threat-intelligence-in/9781492049302/ch01.html
- [3] https://ovalproject.github.io/getting-started/tutorial/

**Comment by pereyra-m** [comment]
### Conclusion

Even when there is a general format like OVAL to represent different and complex situations, it's necessary to define an even more refined format for the scanner to make the detection process as fast as possible.
Also, all the comments and related metadata content will be removed because this information has to be added only when the event is reported.

This new format will be discussed in another issue.
